### PR TITLE
Apply promotion on products

### PR DESCRIPTION
**Pull request by iamnwi**
### API parameters changes ###

Here, I change the API parameters from taking a list of product IDs to a list of product objects(in JSON format) and supposed that every product will have at least a `product_id` field and `price` field.

This API will return the given list of product objects with their updated prices.

### Testing ###
```
Name                  Stmts   Miss  Cover   Missing
---------------------------------------------------
service/__init__.py      20      3    85%   47-51
service/models.py        44      2    95%   41, 47
service/service.py       86      1    99%   41
---------------------------------------------------
TOTAL                   150      6    96%
```

### Pylint ###
service/model.py: 9.77/10
service/service.py: 2.56/10
tests: 6.12/10

### CODE REVIEW

File: service/service.py [hunk 1]
```diff
@@ -68,6 +68,52 @@ def list_promotions():
...
+def apply_a_promotioin(promotion_id):
+    """
+    Apply a promotion on a given set of products together with their prices
+
+    This endpoint will return those given products with their updated price.
+    Products that are not eligible to the given promotion will be returned without any update
+    """
+    app.logger.info('Apply promotion {%s} to products', promotion_id)
+    check_content_type('application/json')
+    data = request.get_json()
```
Comment (reneesun) [inline service/service.py]: What's this request?

File: service/service.py [hunk 2]
```diff
@@ -68,6 +68,52 @@ def list_promotions():
...
+    eligible_ids = [product['product_id'] for product in promotion.products]
+    print(eligible_ids)
+    for product in products:
+        product_id = product['product_id']
+        try:
+            price = float(product['price'])
+        except ValueError as error:
+            raise DataValidationError('The given product prices cannot convert to a float number')
+        if product_id in eligible_ids:
+            product['price'] = price * (promotion.percentage / 100.0)
```
Comment (reneesun) [inline service/service.py]: For the product that not in the eligible_ids, should we add some logging information, e.g.  it cannot apply this promotion code?

Comment (iamnwi) [reply service/service.py]: Sure, we can do this

File: service/service.py (hunk 1 above)
Comment (iamnwi) [reply service/service.py]: This is the HTTP request which including all the pass in parameters, here is the product information

Comment (reneesun) [reply service/service.py]: Got it! Thanks

**Pull request merged**